### PR TITLE
refactor: simplify gas price related code

### PR DESCRIPTION
**Pull request by sergey-melnychuk**
Fixes https://github.com/eqlabs/pathfinder/issues/1306.

Avoid double-`Option` for cached gas price:
- `gas_price: Option<Cached>`
- `Cached::get() -> Option<U256>`

Instead there is an empty state (default) of `Cached` and it is used directly.
Also updated error-messages to avoid confusion about `eth_gasPrice`.

### CODE REVIEW

File: crates/rpc/src/gas_price.rs [hunk 1]
```diff
@@ -1,112 +1,93 @@
...
-                        }
-                    };
-
-                    let now = std::time::Instant::now();
-
-                    let gas_price = gas_price.0.into();
-
-                    let mut g = inner.lock().unwrap_or_else(|e| e.into_inner());
-                    g.latest.replace((now, gas_price));
+        if let Some(gas_price) = self.gas_price().await {
```
Comment (kkovaacs) [inline crates/rpc/src/gas_price.rs:52]: This new code does not seem to do request coalescing anymore. I'd actually prefer keeping the previous logic that guaranteed that there will be at most one pending request at the same time.

File: crates/rpc/src/v02/method/call.rs [hunk 2]
```diff
@@ -2,11 +2,7 @@ use crate::felt::RpcFelt;
 use crate::{context::RpcContext, v03::method::common::base_block_and_pending_for_call};
 use pathfinder_common::{BlockId, CallParam, CallResultValue, ContractAddress, EntryPoint};
 
-crate::error::generate_rpc_error_subset!(
-    CallError: BlockNotFound,
-    ContractNotFound,
-    ContractError
-);
+crate::error::generate_rpc_error_subset!(CallError: BlockNotFound, ContractNotFound, ContractError);
```
Comment (kkovaacs) [inline crates/rpc/src/v02/method/call.rs:5]: This seems to be unrelated -- was it a non-current-stable version of `cargo fmt` that produced this change? Mine does not seem to reformat this line.

File: crates/rpc/src/context.rs [hunk 3]
```diff
@@ -122,7 +122,7 @@ impl RpcContext {
 
     pub fn with_eth_gas_price(self, gas_price: gas_price::Cached) -> Self {
         Self {
-            eth_gas_price: Some(gas_price),
+            eth_gas_price: gas_price,
             ..self
         }
     }
```
Comment (kkovaacs) [inline crates/rpc/src/context.rs:128]: If we remove the `Option` this method no longer makes sense -- there's no way to construct an RpcContext without a gas price source and calling `RpcContext::new()` is equivalent to what `pathfinder/main.rs` does now with `with_eth_gas_price()`.

File: crates/rpc/src/gas_price.rs (hunk 1 above)
Comment (CHr15F0x) [reply crates/rpc/src/gas_price.rs:57]: Shouldn't the outer lock be an async version?

Assuming that the cached value is outdated and the first caller triggered a get_block, the second caller to Cached::get will cause the thread of the second caller to block until get_block is done or times out. Which means all other futures in that second thread will also be blocked.

File: crates/rpc/src/gas_price.rs [hunk 4]
```diff
@@ -1,112 +1,98 @@
...
+                Some(Value { gas_price, updated }) if updated.elapsed() < self.horizon => {
+                    tracing::debug!(?gas_price, from=?updated, "Using cached gas price value");
+                    return Some(gas_price);
+                }
+                _ => {
+                    tracing::debug!("Gas price missing or expired");
                 }
+            },
+            Err(reason) => {
+                tracing::debug!(%reason, "Failed to lock gas price mutex");
```
Comment (kkovaacs) [inline crates/rpc/src/gas_price.rs:53]: Given that we now have the outer lock -- how does this locking failure happen?

Comment (kkovaacs) [reply crates/rpc/src/gas_price.rs:53]: Or to rephrase this: wouldn't having a single async mutex around the actual data and `lock()`-ing that be enough?

Having these two mutexes seems excessive to me.

File: crates/rpc/src/gas_price.rs (hunk 1 above)
Comment (kkovaacs) [reply crates/rpc/src/gas_price.rs:57]: > Shouldn't the outer lock be an async version?

I think it already is a `tokio::sync::Mutex`.

Comment (CHr15F0x) [reply crates/rpc/src/gas_price.rs:57]: > I think it already is a tokio::sync::Mutex.

You're right, my bad.

**Review by CHr15F0x — APPROVED** [review]

File: crates/rpc/src/gas_price.rs [hunk 5]
```diff
@@ -1,112 +1,90 @@
...
-                            return;
-                        }
-                    };
+        let mut guard = self.value.write().await;
 
-                    let now = std::time::Instant::now();
+        // Double-check for cached gas price, to avoid concurrent requests to the gateway.
+        if let Some(Value { gas_price, updated }) = *guard {
+            if updated.elapsed() <= self.horizon {
+                tracing::debug!(?gas_price, from=?updated, "Using cached gas price value");
```
Comment (kkovaacs) [inline crates/rpc/src/gas_price.rs:48]: Can we maybe make this a `trace!()`-level message? These _could_ be useful when diagnosing an error but otherwise would mean we get a lot of messages when running with RUST_LOG=debug during normal operation.

The same goes for all non-error-related messages below.

Comment (CHr15F0x) [reply crates/rpc/src/gas_price.rs:48]: +1

**Review by kkovaacs — APPROVED** [review]
